src/python/tests/test_hal_gene_liftover.py

```diff
@@ -0,0 +1,191 @@
+              'bed_file': 'a2b.one2one.plus.flank1.src.bed', 'flank_length': 1}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 15, 18, '-')],
+              'bed_file': 'a2b.one2one.minus.flank0.src.bed', 'flank_length': 0}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 15, 18, '-')],
+              'bed_file': 'a2b.one2one.minus.flank1.src.bed', 'flank_length': 1}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 22, 25, '+')],
+              'bed_file': 'b2a.one2one.plus.flank0.src.bed', 'flank_length': 0}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 22, 25, '+')],
+              'bed_file': 'b2a.one2one.plus.flank1.src.bed', 'flank_length': 1}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 22, 25, '-')],
+              'bed_file': 'b2a.one2one.minus.flank0.src.bed', 'flank_length': 0}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 22, 25, '-')],
+              'bed_file': 'b2a.one2one.minus.flank1.src.bed', 'flank_length': 1}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 6, 8, '+')],
+              'bed_file': 'a2b.one2many.flank0.src.bed', 'flank_length': 0}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 6, 8, '+')],
+              'bed_file': 'a2b.one2many.flank1.src.bed', 'flank_length': 1}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 26, 32, '+')],
+              'bed_file': 'a2b.inversion.flank0.src.bed', 'flank_length': 0}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 26, 32, '+')],
+              'bed_file': 'a2b.inversion.flank1.src.bed', 'flank_length': 1}, does_not_raise()),
+            ({'regions': [SimpleRegion('chr1', 15, 18, '+')],
+              'bed_file': 'a2b.negative_flank.src.bed', 'flank_length': -1},
+             raises(ValueError, match=r"'flank_length' must be greater than or equal to 0: -1")),
+
+        ]
+    )
+    def test_make_src_region_file(self, kwargs: Dict, expectation: ContextManager,
+                                  tmp_dir: Path) -> None:
+        """Tests :func:`hal_gene_liftover.make_src_region_file()` function.
+
+        Args:
+            kwargs: Named arguments to be passed to the function.
+            expectation: Context manager for the expected exception, i.e. the test will only pass
+                         if that exception is raised. Use :class:`~contextlib.nullcontext` if no
+                         exception is expected.
+            tmp_dir: Unit test temp directory (fixture).
+
+        """
+        with expectation:
+            out_file_path = tmp_dir / kwargs['bed_file']
+            hal_gene_liftover.make_src_region_file(kwargs['regions'], out_file_path,
+                                                   kwargs['flank_length'])
+            ref_file_path = self.ref_file_dir / kwargs['bed_file']
+            assert filecmp.cmp(out_file_path, ref_file_path)
```

Comment:
Same as before regarding `kwargs` and the number of tests (every "ok" test was checking the same code):
```suggestion
    @pytest.mark.parametrize(
        "regions, bed_file, flank_length, expectation",
        [
            ([SimpleRegion('chr1', 15, 18, '+')], 'a2b.one2one.plus.flank0.src.bed', 0, does_not_raise()),
            ([SimpleRegion('chr1', 15, 18, '+')], 'a2b.negative_flank.src.bed', -1,
             raises(ValueError, match=r"'flank_length' must be greater than or equal to 0: -1"))
        ]
    )
    def test_make_src_region_file(self, regions: List[SimpleRegion], bed_file: str, flank_length: int,
                                  expectation: ContextManager, tmp_dir: Path) -> None:
        """Tests :func:`hal_gene_liftover.make_src_region_file()` function.

        Args:
            regions: Regions to write to output file.
            bed_file: Path of BED file to output.
            flank_length: Length of upstream/downstream flanking regions to request.
            expectation: Context manager for the expected exception, i.e. the test will only pass if that
                exception is raised. Use :class:`~contextlib.nullcontext` if no exception is expected.
            tmp_dir: Unit test temp directory (fixture).

        """
        with expectation:
            out_file_path = tmp_dir / bed_file
            hal_gene_liftover.make_src_region_file(regions, out_file_path, flank_length)
            ref_file_path = self.ref_file_dir / bed_file
            assert filecmp.cmp(out_file_path, ref_file_path)
```

Reply:
All points taken, as before. I've create a new suggestion with some changes.
1. Addition of a `chrom_sizes` parameter, for compatibility with the updated version of `make_src_region_file`.
2.  Addition of test cases to check that the input region is within the bounds of the specified chromosome, that the flanked region is being clamped to those bounds, and that the specified chromosome is present in the input.
3. Change of `List[SimpleRegion]` in the `regions` type hint to `Iterable[SimpleRegion]`, to more closely match `hal_gene_liftover.py`.
```suggestion
    @pytest.mark.parametrize(
        "regions, chrom_sizes, bed_file, flank_length, expectation",
        [
            ([SimpleRegion('chr1', 15, 18, '+')], {'chr1': 33}, 'a2b.one2one.plus.flank0.src.bed', 0, does_not_raise()),
            ([SimpleRegion('chr1', 15, 18, '+')], {'chr1': 33}, 'a2b.one2one.plus.flank1.src.bed', 1, does_not_raise()),
            ([SimpleRegion('chr1', 0, 2, '+')], {'chr1': 33}, 'a2b.chrom_start.flank1.src.bed', 1, does_not_raise()),
            ([SimpleRegion('chr1', 31, 33, '+')], {'chr1': 33}, 'a2b.chrom_end.flank1.src.bed', 1, does_not_raise()),
            ([SimpleRegion('chr1', 15, 18, '+')], {'chr1': 33}, 'a2b.negative_flank.src.bed', -1,
             raises(ValueError, match=r"'flank_length' must be greater than or equal to 0: -1")),
            ([SimpleRegion('chrN', 0, 3, '+')], {'chr1': 33}, 'a2b.unknown_chrom.src.bed', 0,
             raises(ValueError, match=r"chromosome ID not found in input file: 'chrN'")),
            ([SimpleRegion('chr1', 31, 34, '+')], {'chr1': 33}, 'a2b.chrom_end.oor.src.bed', 0,
             raises(ValueError, match=r"region end \(34\) must not be greater than chromosome length \(33\)"))
        ]
    )
    def test_make_src_region_file(self, regions: Iterable[SimpleRegion],
                                  chrom_sizes: Mapping[str, int], bed_file: str, flank_length: int,
                                  expectation: ContextManager, tmp_dir: Path) -> None:
        """Tests :func:`hal_gene_liftover.make_src_region_file()` function.

        Args:
            regions: Regions to write to output file.
            chrom_sizes: Dictionary mapping chromosome names to their lengths.
            bed_file: Path of BED file to output.
            flank_length: Length of upstream/downstream flanking regions to request.
            expectation: Context manager for the expected exception, i.e. the test will only pass if that
                exception is raised. Use :class:`~contextlib.nullcontext` if no exception is expected.
            tmp_dir: Unit test temp directory (fixture).

        """
        with expectation:
            out_file_path = tmp_dir / bed_file
            hal_gene_liftover.make_src_region_file(regions, chrom_sizes, out_file_path,
                                                   flank_length)
            ref_file_path = self.ref_file_dir / bed_file
            assert filecmp.cmp(out_file_path, ref_file_path)
```